### PR TITLE
Fix unit tests for WordPress nightly

### PR DESCRIPTION
**Pull request by sjinks**
## Description

The nightly version of WordPress requires `yoast/phpunit-polyfills` to run unit tests.

Without the polyfills, it [throws an error](https://app.circleci.com/pipelines/github/Automattic/vip-go-mu-plugins/6933/workflows/92e8384c-31db-4547-afe6-ba550d8900a4/jobs/49746?invite=true#step-105-62)

> Error: The PHPUnit Polyfills library is a requirement for running the WP test suite.

This PR adds the missing polyfill.

### Fix unit tests in WordPress Nightly

Adds yoast/phpunit-polyfills to composer.{json,lock}

## Checklist

Please make sure the items below have been covered before requesting a review:

- [x] This change works and has been tested locally (or has an appropriate fallback).
- [ ] This change works and has been tested on a Go sandbox.
- [x] This change has relevant unit tests (if applicable).
- [x] This change has relevant documentation additions / updates (if applicable).
- [x] I've created a changelog description that aligns with the provided examples.

## Steps to Test

`php74-build-multisite-nightly` and `php74-build-singlesite-nightly` CI jobs should succeed.

### CODE REVIEW

**Comment by sjinks** [comment]
Not sure why CI jobs don't run :-(

**Comment by rinatkhaziev** [comment]
@sjinks thanks for the PR! 

I'm not sure why the CI is not running, I checked CircleCI and the PR is not there. Can you please try to push a dummy commit (like whitespace or something) and see if it re-triggers it?

I tried to run it locally with the CircleCI CLI tool (`circleci local execute --job php74-build-singlesite-nightly`), and I have good news and bad news. 

The good news is that it dies at a different point, and the bad news is that it still dies :(

Here's a part of the trace:

```
Fatal error: Cannot make static method Yoast\PHPUnitPolyfills\TestCases\TestCase::getProperty() non static in class A8C_Files_Image_Test in /home/circleci/project/tests/files/test-image.php on line 79

Call Stack:
    0.0001     401504   1. {main}() /home/circleci/project/vendor/phpunit/phpunit/phpunit:0
    0.0022     834488   2. PHPUnit\TextUI\Command::main($exit = ???) /home/circleci/project/vendor/phpunit/phpunit/phpunit:61
    0.0022     834600   3. PHPUnit\TextUI\Command->run($argv = [0 => 'vendor/bin/phpunit'], $exit = TRUE) /home/circleci/project/vendor/phpunit/phpunit/src/TextUI/Command.php:162
    0.0022     834600   4. PHPUnit\TextUI\Command->handleArguments($argv = [0 => 'vendor/bin/phpunit']) /home/circleci/project/vendor/phpunit/phpunit/src/TextUI/Command.php:173
    1.2218   42868464   5. PHPUnit\Util\Configuration->getTestSuiteConfiguration($testSuiteFilter = '') /home/circleci/project/vendor/phpunit/phpunit/src/TextUI/Command.php:912
    1.2218   42869048   6. PHPUnit\Util\Configuration->getTestSuite($testSuiteNode = class DOMElement { public $tagName = 'testsuite'; public $schemaTypeInfo = NULL; public $nodeName = 'testsuite'; public $nodeValue = './tests/'; public $nodeType = 1; public $parentNode = '(object value omitted)'; public $childNodes = '(object value omitted)'; public $firstChild = '(object value omitted)'; public $lastChild = '(object value omitted)'; public $previousSibling = NULL; public $nextSibling = NULL; public $attributes = '(object value omitted)'; public $ownerDocument = '(object value omitted)'; public $namespaceURI = NULL; public $prefix = ''; public $localName = 'testsuite'; public $baseURI = '/home/circleci/project/'; public $textContent = './tests/' }, $testSuiteFilter = '') /home/circleci/project/vendor/phpunit/phpunit/src/Util/Configuration.php:996
```

**Comment by sjinks** [comment]
CircleCI still didn't work, but OK, I know what the problem is with the test. There is a trait `AssertAttributeHelper` which defines static `getProperty()` and `getPropertyValues()` methods; `getProperty()` conflicts with `A8C_Files_Image_Test`'s.

I will try to go the slow `./bin/phpunit-docker.sh` route.

File: phpcs.xml.dist
Comment (sjinks) [inline phpcs.xml.dist:52]: Strictly speaking, this is not related to this PR, I only added it to trigger the CI build. I can create a new PR with this fix alone.

File: bin/phpunit-docker.sh
Comment (sjinks) [inline bin/phpunit-docker.sh:69]: This adds support for WordPress Nightly. The problem with the current configuration is that if you are using the "global" phpunit (in our case, the one provided by the `ghcr.io/automattic/phpunit-docker/phpunit`), the polyfills need to be installed globally as well.

Since we have installed `yoast/phpunit-polyfills` locally, we have to override the entry point script for the container.
`docker history` says that the original entry point was `/usr/local/bin/phpunit`, therefore we can painlessly update it to `/app/vendor/bin/phpunit` provided that the file exists.

File: tests/files/test-image.php
Comment (sjinks) [inline tests/files/test-image.php:79]: The `AssertAttributeHelper` trait from `yoast/phpunit-polyfills` adds two static methods: `getProperty()` and `getPropertyValue()`. The first one conflicts with the definition of our `getProperty()`. The implementation of both methods is incompatible; this makes it very difficult to fix `test-image.php` to use the pollyfilled methods.

I, therefore, decided to rename the methods using the snake case (this seems to be in line with the coding standards).

File: tests/files/test-image.php
Comment (sjinks) [inline tests/files/test-image.php:318]: The more robust approach would be to subclass the `Automattic\VIP\Files\Image`, and provide public access to `set_width_height()` and `get_resized_filename()` methods. This would eliminate the need to use the Reflection API and improve the readability of the code (not to mention that it would make static analysis tools happier).

**Comment by sjinks** [comment]
@rinatkhaziev I am running the full test suite with `./bin/phpunit-docker.sh --wp nightly`, this is going to take several hours on my notebook. I will check the results in my morning, and will fix the failing tests, if any,

**Comment by sjinks** [comment]
Passed `./bin/phpunit-docker.sh --wp nightly` and `./bin/phpunit-docker.sh --wp 5.8` tests. Still, it would be nice if the CI can run for this PR :slightly_smiling_face:

**Comment by jblz** [comment]
👋 I think the CI isn't running b/c this PR is from an untrusted fork. See https://circleci.com/blog/triggering-trusted-ci-jobs-on-untrusted-forks/

**Comment by pschoffer** [comment]
Thank you @sjinks for helping on this!

I branched out from your branch and pushed it to get the circleCI [run](https://app.circleci.com/pipelines/github/Automattic/vip-go-mu-plugins?branch=fix%2F2367_test_run)

However, the tests started to fall now, due to an issue unrelated to your change. I will investigate.

Also though, @pauarge was looking into the failing tests, and I believe he mentioned there is a core fix coming up as the recent change broke test pipelines for several related projects.

**Comment by sjinks** [comment]
@pschoffer if this helps, there is no `libgcc-8-dev` in Debian 11 (Bullseye), it has only `libgcc-9-dev` and `libgcc-10-dev`. `libgcc-8-dev` is present in Debian 10 (Buster) though.

I think the issue is that CircleCI uses images like `circleci/php:7.4-node` image when you probably want `circleci/php:7.4-buster-node` instead.

**Comment by sjinks** [comment]
Possible fix:

```diff
diff --git a/.circleci/config.yml b/.circleci/config.yml
index 52698ebb1..2446a7de5 100644
--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -68,7 +68,7 @@ job-references:
   install_dependencies: &install_dependencies
     name: "Install Dependencies"
     command: |
-      sudo apt-get update && sudo apt-get install subversion libgcc-8-dev default-mysql-client libpng-dev
+      sudo apt-get update && sudo apt-get install subversion libgcc-10-dev default-mysql-client libpng-dev
       sudo -E docker-php-ext-install mysqli gd
       npm ci
       composer install -n
@@ -76,7 +76,7 @@ job-references:
   install_core_tests_dependencies: &install_core_tests_dependencies
     name: "Install Dependencies"
     command: |
-      sudo apt-get update && sudo apt-get -y install subversion libgcc-8-dev default-mysql-client nodejs npm
+      sudo apt-get update && sudo apt-get -y install subversion libgcc-10-dev default-mysql-client nodejs npm
       sudo npm install npm@latest -g
       composer install -n
```

I will try to create a PR to test this in CircleCI (not sure if it works this time). I did some limited testing (circleci CLI does not work for me, so I ran install commands manually) on all images mentioned in `config.yml`.

**Comment by sjinks** [comment]
I have created a PR: #2371 but it does not trigger CircleCI :-(

**Comment by pauarge** [comment]
Thanks @sjinks for your contribution.

As @pschoffer mentions, we were internally waiting for some changes in Core. This is failing because the Core team are doing some changes ahead of 5.9, which is still not released. Concretely, they are have a couple of PRs that are directly related with this that are still under review: https://github.com/WordPress/wordpress-develop/pull/1587 and https://github.com/WordPress/wordpress-develop/pull/1588.

File: tests/files/test-image.php
Comment (pauarge) [reply tests/files/test-image.php:318]: That sounds great, we could leave it for another PR.

File: tests/files/test-image.php
Comment (pauarge) [reply tests/files/test-image.php:79]: It makes sense. Could a brief explanation of this be added in the method's comment? Something like `Renaming to get_property to avoid conflict with polyfills`?

File: phpcs.xml.dist
Comment (pauarge) [reply phpcs.xml.dist:52]: If this is not related to the PR, I'd seaparate it on its own to better track it down the line.

**Review by pauarge — CHANGES_REQUESTED** [review]
Thanks for your work, this looks very good. I've left a couple of small improvements that we could do. 

As I mentioned in the comment above, this code might get simplified once Core merges a their changes, so we might have to revisit it. I'm happy to merge it now, since it brings back all our tests to green.

**Comment by sjinks** [comment]
OK cool — I will address the CR comments now.

File: phpcs.xml.dist
Comment (sjinks) [reply phpcs.xml.dist:52]: Moved this into a separate PR: #2373

Will drop this commit after rebase.

File: tests/files/test-image.php
Comment (sjinks) [reply tests/files/test-image.php:84]: Done in 1c7eadd6f

File: phpcs.xml.dist
Comment (sjinks) [reply phpcs.xml.dist]: Done.

**Review by pauarge — APPROVED** [review]

**Pull request merged**
